src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -99,47 +99,72 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Diagno
+                ConstantPatternSyntax p => BindConstantPatternWithFallbackToTypePattern(p, inputType, hasErrors, diagnostics),
+                RecursivePatternSyntax p => BindRecursivePattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                VarPatternSyntax p => BindVarPattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                ParenthesizedPatternSyntax p => BindPattern(p.Pattern, inputType, inputValEscape, hasErrors, diagnostics),
+                BinaryPatternSyntax p => BindBinaryPattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                UnaryPatternSyntax p => BindUnaryPattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                RelationalPatternSyntax p => BindRelationalPattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                TypePatternSyntax p => BindTypePatternWithFallbackToConstantPattern(p, inputType, inputValEscape, hasErrors, diagnostics),
+                _ => throw ExceptionUtilities.UnexpectedValue(node.Kind()),
+            };
         }
 
         private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol inputType)
         {
             return new BoundDiscardPattern(node, inputType);
         }
 
-        private BoundConstantPattern BindConstantPattern(
+        private BoundPattern BindConstantPatternWithFallbackToTypePattern(
             ConstantPatternSyntax node,
             TypeSymbol inputType,
             bool hasErrors,
             DiagnosticBag diagnostics)
         {
-            SyntaxNode innerExpression = node.Expression.SkipParens();
+            return BindConstantPatternWithFallbackToTypePattern(node, node.Expression, inputType, hasErrors, diagnostics);
+        }
+
+        internal BoundPattern BindConstantPatternWithFallbackToTypePattern(
+            SyntaxNode node,
+            ExpressionSyntax expression,
+            TypeSymbol inputType,
+            bool hasErrors,
+            DiagnosticBag diagnostics)
+        {
+            ExpressionSyntax innerExpression = expression.SkipParens();
             if (innerExpression.Kind() == SyntaxKind.DefaultLiteralExpression)
             {
                 diagnostics.Add(ErrorCode.ERR_DefaultPattern, innerExpression.Location);
                 hasErrors = true;
             }
 
-            return BindConstantPattern(node, inputType, node.Expression, hasErrors, diagnostics, out _);
+            var constantDiagnostics = DiagnosticBag.GetInstance();
+            var constantPattern = BindConstantPattern(node, inputType, expression, hasErrors, constantDiagnostics, out bool wasExpression);
+
+            // fall back to a type pattern if it fails as a constant pattern.
+            if (!wasExpression && !IsUnderscore(innerExpression) &&
+                ((CSharpParseOptions)node.SyntaxTree.Options).IsFeatureEnabled(MessageID.IDS_FeatureTypePattern))
```

Comment:
We should attempt to bind a type pattern here regardless of the language version, and then issue a LangVersion diagnostic if the feature is not enabled.